### PR TITLE
Fix fast up to date

### LINKED ISSUE #4665: Consuming projects are not updated when a CopyToOutputDirectory item is updated in a project with no project references.
**Visual Studio Version**: 2017 - 15.9.10

Also reproduced in 2019 RC.4 - 16.0.0

**Summary:**

Fast up-to-date checks will consider a project to be up-to-date even when a project it references is not up to date due to a stale `CopyToOutputDirectory` item.

(I've found some workarounds which are [described here](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated#workarounds).)

**Steps to Reproduce**:

1. [Clone this repository](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated).
    * The demonstration repository contains details on how the projects are set up. The most important thing to know is that `TextFileA.txt` and `TextFileB.txt` are copied to the output directory with `PreserveNewest` and the program prints their contents when it runs.
    * The repository also contains more details on my findings about the source of the issue.
    * The steps from the repository for reproducing the problem are reproduced below for posterity.
1. Open the solution in Visual Studio.
2. Select `TestProgram` as the startup project.
3. Press F5 to start the program. The output is as follows:
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
4. Close the program and add a message `TextFileA.txt` and press F5 to start the program again. The output includes your message:
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty! Hello, world!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
5. Close the program and add a message to `TextFileB.txt` instead, press F5 to start the program again. **The output will not include your message:**
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty! Hello, world!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
6. Observe that the file is updated in `LibraryB`'s output (`LibraryB\bin\Debug\netstandard2.0\TextFileB.txt`), but not in `TestProgram`'s output (`TestProgram\bin\Debug\netcoreapp2.2\TextFileB.txt`).

**Expected Behavior**:

`TextFileB.txt` is updated in `TestProgram`'s output folder when it is changed.

**Actual Behavior**:

`TextFileB.txt` is not updated.

**User Impact**:

Ironically I actually found this bug while debugging a similar issue with a legacy-style C# project. Either way, the symptoms are the same even though it appears the cause is not.

This issue is very confusing because you can see that `LibraryB` was built by Visual Studiom, but projects depending on it directly/indirectly were not. The copied file will be present in the output of `LibraryB`, but not in `TestProgram`. You can rebuild the solution or make dummy changes in `TestProgram` to force things to update, but this is fairly annoying to remember to do. In our case, the copied file was a native library, so it was like changes to the native library weren't working.

**Workarounds**:

Besides rebuilding, I've found two workarounds to this issue:
1. Add a dummy reference to the library.
2. Create the copy marker yourself using MSBuild.

Both are [described in further detail in my demonstration repository](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated#workarounds).

### PR DESCRIPTION
**Pull request by ocallesp**
- Added the capability to check items in CopyToOutputDirectory from project references.
- Update unit tests

This implementation fixes #4665

### CODE REVIEW

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs [hunk 1]
```diff
@@ -1031,12 +1038,18 @@ public async Task IsUpToDateAsync_False_CopyToOutputDirectorySourceIsNewerThanDe
...
         public async Task IsUpToDateAsync_False_CopyToOutputDirectorySourceDoesNotExist()
         {
-            var sourceSnapshot = new Dictionary<string, IProjectRuleSnapshotModel>
+            var projectSnapshot = new Dictionary<string, IProjectRuleSnapshotModel>
             {
-                [Content.SchemaName] = ItemWithMetadata("Item1", "CopyToOutputDirectory", "PreserveNewest")
+                [CopyToOutputDirectoryItem.SchemaName] = new IProjectRuleSnapshotModel
+                {
+                    Items = ImmutableStringDictionary<IImmutableDictionary<string, string>>.EmptyOrdinal
+                        .Add("C:\\Dev\\Solution\\Project\\Item1", ImmutableStringDictionary<string>.EmptyOrdinal
```
Comment (jjmew) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs]: >C [](start = 30, length = 1)

Do we always have access to c:? i was just thinking that in CI we might not.

Comment (ocallesp) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs]: In this implementation we check CopyToOutputDirectory items from jointRuleUpdate (projectSnapshot) which needs to receive a[ rooted path](https://github.com/dotnet/project-system/blob/482f68977fd3ade7270cb3b8995565b926d05b73/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs#L522-L529). 

In the unit test this rooted path might represent a file from  project reference at path "C:\\Dev\\Solution\\Project\\Item1" marked as PreserveNewest.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 2]
```diff
@@ -520,47 +517,40 @@ private bool CheckCopiedOutputFiles(Log log, in TimestampCache timestampCache, S
...
+        private bool CheckPreserveNewestCopyToOutputDirectoryFiles(Log log, in TimestampCache timestampCache, State state)
         {
-            IEnumerable<(string path, string? link, CopyType copyType)> items = state.ItemsByItemType.SelectMany(kvp => kvp.Value).Where(item => item.copyType == CopyType.CopyIfNewer);
+            IEnumerable<KeyValuePair<string, (string targetPath, CopyType copyType)>> preserveNewestItems
+                = state.CopyToOutputDirectoryItems.Where(item => item.Value.copyType == CopyType.PreserveNewest);
 
             string outputFullPath = Path.Combine(state.MSBuildProjectDirectory, state.OutputRelativeOrFullPath);
 
-            foreach ((string path, string? link, _) in items)
+            foreach ((string sourcePath, (string targetFilename, _)) in preserveNewestItems)
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Nit: We often use "fullPath" to indicate this a rooted path, "sourceFullPath".

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs (hunk 1 above)
Comment (davkean) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs]: These are just made up paths, they don't need to exist.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs (hunk 2 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: I will change the variable name to sourceFullPath to make it clear.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 3]
```diff
@@ -31,21 +31,18 @@ internal sealed partial class BuildUpToDateCheck : OnceInitializedOnceDisposedUn
         private const string DefaultSetName = "";
         private static readonly StringComparer s_setNameComparer = StringComparers.ItemNames;
 
-        private static ImmutableHashSet<string> ReferenceSchemas => ImmutableStringHashSet.EmptyOrdinal
+        // This order is used everywhere, i.e., in the documentation.
+        // Do not change the order.
```
[the remark this answers is not in the document]
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: I will delete this comment

**Pull request merged**